invokeai/app/services/batch_manager.py

```diff
@@ -0,0 +1,206 @@
+    def run_batch_process(self, batch_id: str) -> None:
+        """Runs a batch process"""
+        pass
+
+    @abstractmethod
+    def cancel_batch_process(self, batch_process_id: str) -> None:
+        """Cancels a batch process"""
+        pass
+
+    @abstractmethod
+    def get_batch(self, batch_id: str) -> BatchProcessResponse:
+        """Gets a batch process"""
+        pass
+
+    @abstractmethod
+    def get_batch_processes(self) -> list[BatchProcessResponse]:
+        """Gets all batch processes"""
+        pass
+
+    @abstractmethod
+    def get_incomplete_batch_processes(self) -> list[BatchProcessResponse]:
+        """Gets all incomplete batch processes"""
+        pass
+
+    @abstractmethod
+    def get_sessions(self, batch_id: str) -> list[BatchSession]:
+        """Gets the sessions associated with a batch"""
+        pass
+
+
+class BatchManager(BatchManagerBase):
+    """Responsible for managing currently running and scheduled batch jobs"""
+
+    __invoker: Invoker
+    __batch_process_storage: BatchProcessStorageBase
+
+    def __init__(self, batch_process_storage: BatchProcessStorageBase) -> None:
+        super().__init__()
+        self.__batch_process_storage = batch_process_storage
+
+    def start(self, invoker: Invoker) -> None:
+        self.__invoker = invoker
+        local_handler.register(event_name=EventServiceBase.session_event, _func=self.on_event)
+
+    async def on_event(self, event: Event):
+        event_name = event[1]["event"]
+
+        match event_name:
+            case "graph_execution_state_complete":
+                await self._process(event, False)
+            case "invocation_error":
+                await self._process(event, True)
+
+        return event
+
+    async def _process(self, event: Event, err: bool) -> None:
+        data = event[1]["data"]
+        try:
+            batch_session = self.__batch_process_storage.get_session_by_session_id(data["graph_execution_state_id"])
+        except BatchSessionNotFoundException:
+            return None
+        changes = BatchSessionChanges(state="error" if err else "completed")
+        batch_session = self.__batch_process_storage.update_session_state(
+            batch_session.batch_id,
+            batch_session.session_id,
+            changes,
+        )
+        batch_process = self.__batch_process_storage.get(batch_session.batch_id)
+        if not batch_process.canceled:
+            self.run_batch_process(batch_process.batch_id)
+
+    def _create_batch_session(self, batch_process: BatchProcess, batch_indices: tuple[int]) -> GraphExecutionState:
+        graph = batch_process.graph.copy(deep=True)
+        batch = batch_process.batch
+        g = graph.nx_graph_flat()
+        sorted_nodes = nx.topological_sort(g)
+        for npath in sorted_nodes:
+            node = graph.get_node(npath)
+            for index, bdl in enumerate(batch.data):
+                relevant_bd = [bd for bd in bdl if bd.node_id in node.id]
```

Comment:
I don't think this will work. `node.id` will just be the local path to a node, whereas the node path would include sub-paths for subgraphs. If you're using full node paths for batch data (which it seems like you should), then you should just loop your batch data and get relevant nodes by node path. Then you don't need to do the `nx_graph_flat` call, which isn't super performant (and you're also only looping over data you want to replace instead of the entire graph).

Reply:
Yep, done. Works with subgraphs. Thanks.